src/parser.rs

```diff
@@ -212,26 +212,30 @@ impl<'a> Parser<'a> {
 
     pub fn path(self) -> Result<'a, String> {
         let (path, parser) = self.word()?;
-        let mut path = Path::new(path);
 
-        // For Linux and BSD and MacOS.
+        if path == "" {
+            return Ok((String::from(""), parser));
+        }
+
+        let mut path = PathBuf::from(path);
+
+        // Linux and BSD and MacOS use ~ to infer the home directory of a given user
         if cfg!(unix) && path.starts_with("~") {
             if let Some(base_dirs) = dirs::BaseDirs::new() {
-                let home = base_dirs.home_dir().to_str().unwrap().to_owned();
-
-                if path == Path::new("~") {
-                    return Ok((home, parser));
-                }
 
-                path = &path.strip_prefix("~").unwrap();
-                return Ok((
-                    Path::new(&home).join(path).to_str().unwrap().to_owned(),
-                    parser,
-                ));
+                let suffix = path.strip_prefix("~").unwrap();
```

Comment:
I think we can avoid this `unwrap` by using `strip_prefix` instead of `path.starts_with("~")` above, to both check for "~" and remove it if exists.

Reply:
Oh I see that using `unwrap()` is bad style in Rust, I take it?

I suppose I could use `map()` or something to overload an if-let-statement or something. Since I am somewhat unfamiliar with Rust, if you don't mind me asking, what would you consider to be the proper Rust way of solving this? :smile:

Reply:
Yeah, because it can cause a *panic* and we don't want that :smile: 

In this case you could do something like:

```rust
if let Ok(path) = path.strip_prefix("~") {
     // This means `path` had ~ in its prefix, and now it doesn't anymore
} else {
     // Path didn't have strip prefix, so `path` is the original path in here
}
```

Reply:
You can always also use the `?` syntax to bubble up errors: https://doc.rust-lang.org/std/result/#the-question-mark-operator- since the return type of this function is also `Result`. And if the error types don't match, you can do:

```rust
let path = path.to_str().map_err(|e| Err(Error::new(e)))?;
```
For example, which converts the error from `to_str()` into our own error type, and returns the value in the `Ok` if there was no error.

Reply:
Much appreciated for the help :smiley:

That said, I think `to_str()` yields a type `Option` so I'm not sure there is a `map_err()` function for that type, though I think perhaps there is one for type `Result`. 

In any event, I've rewritten that function using your advice from above and removed the unwraps in the home directory section and replaced the final unwrap with a match statement. 

Let me know what you think.